File: Day180_2025/Q180_GFG.java
class Q180_GFG {
    private boolean isFeasible(int[] arr, int k, int maxSum) {
        int sum = 0;
        int count = 1;
        for (int num : arr) {
            sum += num;
            if (sum > maxSum) {
                sum = num;
                count++;
                if (count > k) {
                    return false;
                }
            }
        }
        return true;
    }

    public int splitArray(int[] arr, int k) {
        int left = Integer.MIN_VALUE;
        int right = 0;
        for (int num : arr) {
            left = Math.max(left, num);
            right += num;
        }

        while (left < right) {
            int mid = left + (right - left) / 2;
            if (isFeasible(arr, k, mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }
}

File: Day180_2025/Q180_GFG.py
class Solution:
    def isFeasible(self, arr, k, maxSum):
        count = 1
        total = 0
        for num in arr:
            total += num
            if total > maxSum:
                total = num
                count += 1
                if count > k:
                    return False
        return True

    def splitArray(self, arr, k):
        left, right = max(arr), sum(arr)
        while left < right:
            mid = (left + right) // 2
            if self.isFeasible(arr, k, mid):
                right = mid
            else:
                left = mid + 1
        return left
